Fix confusion matrix type

edge7 opened the pull request:
Hello,

this is an attempt to show and fix the issue I opened:
#20574

fchollet commented:
Thanks for the PR. Unfortunately using `float64` will be incompatible with e.g. the JAX backend.

edge7 commented:
ok, in that case, it might suffice to check the backend and set float64 just in case of torch/tf. However, the results might be wrong with JAX float 32. 
Alternatively, the entire thing (the confusion matrix) can be done with int64, without issues

fchollet commented:
What about int32? Is that enough?

On Tue, Dec 3, 2024, 1:38 PM Enrico ***@***.***> wrote:

> ok, in that case, it might suffice to check the backend and set float64
> just in case of torch/tf. However, the results might be wrong with JAX
> float 32.
> Alternatively, the entire thing (the confusion matrix) can be done with
> int64, without issues
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keras-team/keras/pull/20584#issuecomment-2515602079>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFNM32XICQSJJLSG2YEODD2DYQEPAVCNFSM6AAAAABS6S6VXOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJVGYYDEMBXHE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

edge7 commented:
int32 is good too.
But it breaks a good amount of existing tests.

edge7 commented:
By the way, I suggest caution with *32 as with high resolution images and big dataset the numbers can explode very quickly

fchollet commented:
> The weights should probably also be just int32, which makes sense IMHO

Indeed, can we just do that then?

edge7 commented:
Ok. I'll try to modify the PR (will be much larger as some existing tests rely on float weights)

edge7 commented:
I switched to int32 as the default, which resolved the main issue. I also made a small adjustment to the result function to address a complaint related to the epsilon (which is float). Additionally, I fixed a couple of existing tests and streamlined my own test to make it lighter and avoid OOM problems.

edge7 commented:
Hold on with this one as I am trying to run a training with the changes and I get:

`Trying to access resource mean_io_u_1/total_confusion_matrix/600 (defined @ /anaconda/envs/keras3_jax/lib/python3.10/site-packages/tensorflow/python/ops/gen_resource_variable_ops.py:1267) located in device /job:localhost/replica:0/task:0/device:CPU:0 from device /job:localhost/replica:0/task:0/device:GPU:0`

edge7 commented:
with int64, training starts and the metric looks ok (Tf 2.18 for training).
Also with int64, tests work fine in JAX and Torch.

edge7 commented:
Training with JAX also works with int64, but warnings are raised:

```
/anaconda/envs/keras3_jax/lib/python3.10/site-packages/keras/src/backend/jax/core.py:77: UserWarning: Explicitly requested dtype int64 requested in asarray is not available, and will be truncated to dtype int32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. See https://github.com/google/jax#current-gotchas for more.
  return jnp.asarray(x, dtype=dtype)
```
The metric looks fine anyway, as the truncation does not affect the result. 

To sum up:

- Float 32 is not good enough
- Int 32 is good for representing big integers, but cannot be placed in GPU with TF
- Int 64 works for TF GPU, but raise warning in JAX (but still working)
- Not tested Torch training

james77777778 commented:
Hey @edge7 
It is well-known that we need to use int64 for tensorflow to work on GPUs. We can use a "magic" `int` to let the backend decide whether to use int32 (for jax, torch) or int64 (for tensorflow).
Please refer to https://github.com/keras-team/keras/blob/master/keras/src/backend/common/dtypes.py#L49

You should be able to implement it as follows:
```python
super().__init__(name=name, dtype=dtype or int)
```

fchollet commented on keras/src/metrics/iou_metrics_test.py:
```diff
@@ -64,7 +65,9 @@ def test_multi_dim_input(self):
 
-        obj = metrics.IoU(num_classes=2, target_class_ids=[0, 1])
+        obj = metrics.IoU(
+            num_classes=2, target_class_ids=[0, 1], dtype="float32"
```
Why was it necessary to change the dtypes in the unit tests?

edge7 replied:
float32 is essential for tests involving float-format weights and Binary IoU calculations, as casting problems can result in all predictions being set to zero. This is somewhat inconvenient, as float would typically be the preferred choice, but it becomes unreliable for values exceeding 2^24

edge7 replied:
I'm unsure about the best way to proceed. The issue is clear, but the optimal approach isn't straightforward. Using float64 would be a quick fix, but JAX doesn't handle it well (even though it can be configured to use it, I’m not sure if there are any side effects). Using int32/int64 is a decent alternative, but given the current implementation, passing weights as floats wouldn't make sense in that case. To address this, a warning could be issued, or an exception could be raised, with clear documentation explaining the limitation. Additionally, the IoU binary calculation should be slightly adjusted to evaluate >= threshold in float format regardless of the data type, and only cast to the appropriate type afterward.
